### ISSUE TITLE
提问。使用lora调参14b-chat的模型，为什么显存占用没有很明显的减少？

### ISSUE THREAD

**Issue by ezekielqu** [QwenLM/Qwen#761]
使用的设备：A100 * 8
cuda版本：11.7
环境：

- accelerate                    0.24.1
- annotated-types               0.6.0
- Brotli                        1.0.9
- certifi                       2023.11.17
- cffi                          1.16.0
- chardet                       5.2.0
- charset-normalizer            2.0.4
- cmake                         3.25.0
- cryptography                  41.0.3
- deepspeed                     0.12.4
- dropout-layer-norm            0.1
- einops                        0.7.0
- filelock                      3.13.1
- flash-attn                    2.3.6
- fsspec                        2023.10.0
- gmpy2                         2.1.2
- hjson                         3.1.0
- huggingface-hub               0.19.4
- idna                          3.4
- Jinja2                        3.1.2
- lit                           15.0.7
- MarkupSafe                    2.1.3
- mkl-fft                       1.3.8
- mkl-random                    1.2.4
- mkl-service                   2.4.0
- mpmath                        1.3.0
- networkx                      3.2.1
- ninja                         1.11.1.1
- numpy                         1.26.0
- packaging                     23.2
- peft                          0.6.2
- Pillow                        10.0.1
- pip                           23.3.1
- psutil                        5.9.6
- py-cpuinfo                    9.0.0
- pycparser                     2.21
- pydantic                      2.5.2
- pydantic_core                 2.14.5
- pynvml                        11.5.0
- pyOpenSSL                     23.2.0
- PySocks                       1.7.1
- PyYAML                        6.0.1
- regex                         2023.10.3
- requests                      2.31.0
- rotary-emb                    0.1
- safetensors                   0.4.1
- scipy                         1.11.4
- setuptools                    68.0.0
- sympy                         1.12
- tiktoken                      0.5.1
- tokenizers                    0.13.3
- torch                         2.0.1+cu118
- torchaudio                    2.1.1
- torchvision                   0.15.2+cu117
- tqdm                          4.66.1
- transformers                  4.32.0
- transformers-stream-generator 0.0.4
- triton                        2.0.0
- typing_extensions             4.8.0
- urllib3                       1.26.18
- wheel                         0.41.2

进行全参数微调时使用的参数：
```bash
torchrun $DISTRIBUTED_ARGS finetune.py \
    --model_name_or_path $MODEL \
    --data_path $DATA \
    --bf16 True \
    --output_dir output_qwen1206 \
    --num_train_epochs 5 \
    --per_device_train_batch_size 2 \
    --per_device_eval_batch_size 1 \
    --gradient_accumulation_steps 16 \
    --evaluation_strategy "no" \
    --save_strategy "steps" \
    --save_steps 1000 \
    --save_total_limit 10 \
    --learning_rate 1e-5 \
    --weight_decay 0.1 \
    --adam_beta2 0.95 \
    --warmup_ratio 0.01 \
    --lr_scheduler_type "cosine" \
    --logging_steps 1 \
    --report_to "none" \
    --model_max_length 8192 \
    --gradient_checkpointing True \
    --lazy_preprocess True \
    --deepspeed finetune/ds_config_zero3.json
```
训练时显存占用，每张卡约65g
![image](https://github.com/QwenLM/Qwen/assets/69846604/2394cf7e-8d4f-4088-b1e9-ea698d05f530)

使用lora时使用的参数：
```bash
torchrun $DISTRIBUTED_ARGS finetune.py \
    --model_name_or_path $MODEL \
    --data_path $DATA \
    --bf16 True \
    --output_dir output_qwen_l_1207 \
    --num_train_epochs 5 \
    --per_device_train_batch_size 2 \
    --per_device_eval_batch_size 1 \
    --gradient_accumulation_steps 8 \
    --evaluation_strategy "no" \
    --save_strategy "steps" \
    --save_steps 1000 \
    --save_total_limit 10 \
    --learning_rate 3e-4 \
    --weight_decay 0.1 \
    --adam_beta2 0.95 \
    --warmup_ratio 0.01 \
    --lr_scheduler_type "cosine" \
    --logging_steps 1 \
    --report_to "none" \
    --model_max_length 8192 \
    --lazy_preprocess True \
    --use_lora \
    --gradient_checkpointing \
    --deepspeed finetune/ds_config_zero2.json
```
训练时显存占用，每张卡约55g
![image](https://github.com/QwenLM/Qwen/assets/69846604/5a28c3d8-edb4-4fbd-a535-e0fd5a1ea5e3)

两次训练使用同样的数据集，预期的训练时间也都是大约12个小时，lora也没有显著减少时间或显存占用。请问一下这个是什么原因呢？

**Comment by jklj077** [comment]
LoRA配置下的显存占用好像跟README里我们提供的参考差不多，这个应该符合预期。

另外您提供的脚本里，LoRA开了ZeRO2，全量微调开了ZeRO3哈，显存占用不太能比哈，ZeRO3显存占用本来就小些。

**Comment by ezekielqu** [comment]
> LoRA配置下的显存占用好像跟README里我们提供的参考差不多，这个应该符合预期。
> 
> 另外您提供的脚本里，LoRA开了ZeRO2，全量微调开了ZeRO3哈，显存占用不太能比哈，ZeRO3显存占用本来就小些。

非常感谢您的解答。把lora训练的脚本也改成zero3，显存占用下降到了大约每张卡35g。
还有一个问题想请教一下，对于目前我的机器来说，全量微调和lora都是使用的流水线并行的方式么？因为在训练的log中打印，lora训练的参数量只用原模型的1.5%左右，虽然设置zero3会大幅下降，但是显存占有还是有全量微调的一半。提供的参考是使用单卡训练的情况下，但是我现在是8卡，lora的显存下降就没有想的巨大了。

**Comment by jklj077** [comment]
> 全量微调和lora都是使用的流水线并行的方式么

当前的微调脚本支持的都是数据并行（DeepSpeed ZeRO和FSDP技术意义上都是数据并行，多卡间传的主要是参数/梯度，不是隐状态），是不支持模型并行/流水线并行的。LoRA本身需要梯度的参数没那么多，ZeRO切分优化器、梯度到多卡上的方案平均每张卡也节省不了太多显存。单卡显存能运行的话，可以单卡运行；多卡运行也可以充分的利用显存资源，比如batchsize大一些。
